### ISSUE TITLE
How to handle prefixed dependencies also used by dev-dependencies

### ISSUE THREAD

**Issue by justlevine** [BrianHenryIE/strauss#81]
I've been struggling to understand the best way to handle prefixing a dependency that is in use by both the project itself and some of the project dev dependencies. **I am unsure if this is a bug or a usage issue.**

For example, over in this PR for [axewp/wp-graphql-headless-login](https://github.com/AxeWP/wp-graphql-headless-login/pull/102/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34), we're namespacing several production dependencies that require `guzzlehttp/guzzle`. However, that package is also required by our dev-deps `codeception/module-phpbrowser`, which fails since it doesn't know to use the namespaced version of the classes, causing errors like:
 `PHP Fatal error:  Uncaught Error: Class "GuzzleHttp\HandlerStack" not found in /var/www/html/wp-content/plugins/wp-graphql-headless-login/vendor/codeception/module-phpbrowser/src/Codeception/Lib/Connector/Guzzle.php:342` [here](https://github.com/AxeWP/wp-graphql-headless-login/actions/runs/6936758734/job/18869517940?pr=102).
 
 ### What I've tried
 
 - `delete_vendor_packages` doesn't seem to prefix the usage of deps in non-namespaced packages, nor does it happen if the `target_directory` is set to `vendor`.
 - `exclude_from_copy` and `exclude_from_prefix` don't make sense here, since we explicitly want the package to be namespaced to prevent conflicts with other plugins.
 - Also tried setting `delete_vendor_packages` to `false`, and to conditionally require `vendor/autoload.php`, however I couldnt find an option for strauss to copy the `PSR-4` autoloader over to `vendor-prefixed/autoload.php`, and since `delete_vendor_packages` doesn't seem to be configurable on a per-package level that means our distribution zip needs to include duplicates of all namespaced dependencies.
  - It doesnt seem like `config.strauss` provides a way to have separate commands for `composer install --no-dev` (e.g. delete copy from vendor) and `composer install` (e.g. _dont_ delete copy from vendor), which would let me use both the strauss and composer autoloaders and still keep the duplicates out of the production build.
  
 Is this a bug, known limitation, or just something wrong with how I'm approaching strauss?

**Comment by defunctl** [comment]
I just ran into this as well, with a prefixed Symfony library used by both the Codeception tests library and in our code base.

**Comment by BrianHenryIE** [comment]
I'm not sure why you're ever getting a PHP error – since your project code should be using the copied + prefixed class and the original should still exist in your vendor directory. I use WP Browser/Codeception pretty widely and haven't had a problem. 

I see why it's a problem that the unprefixed package still remains after `composer install --no-dev`. 

I'll break up the tool so commands can be run individually. 

I'm already working on that as part of a feature to prefix the call sites in project files. Which would also give the option of doing all development using the true namespaces and only prefixing as a build step. https://github.com/BrianHenryIE/strauss/commit/2bd1b3f79c01b8d80595ceca383971ad7f5ea9c4

Another approach which might work (but I don't currently plan to do) would be for Strauss to have a `class_alias()` map which could be written on each run and loaded by a `files` autoloader.

**Comment by justlevine** [comment]
> I'm not sure why you're ever getting a PHP error – since your project code should be using the copied + prefixed class and the original should still exist in your vendor directory. I use WP Browser/Codeception pretty widely and haven't had a problem.

The PHP error occurs when the namespaces dependency (GuzzleHttp/Guzzle) is deleted/replaced in situ, since the dev-dependency (Codeception) needs the nonprefixed copy.

If the original is _not_ deleted, then I need to bundle both the original (for use by the dev deps) and the prefixed dependency (for use by project code) in production, since I need the PSR-4 autoloader from `composer.json`.

And since `delete_vendor_packages` is all-or-nothing, that means I need to include duplicates of all the renamed packages as well.

Hope this clarify things. As far as possible solutions go,  Im over my head and happily defer to what you think the ideal mechanism would be. Will note that an additional possible solve would be to have Strauss provide the option to bring over the composer.json `"autoload"` config so a production build could solely rely on `vendor-prefixed/autoload.php`, but no idea if that's a better/worse pattern than what you described.

**Comment by defunctl** [comment]
Same deal for me, with `delete_vendor_packages`: true, I get:

```
Fatal error: Uncaught Error: Interface "Symfony\Contracts\Service\ResetInterface" not found in /var/www/html/wp-content/plugins/project/vendor/symfony/console/Application.php on line 62
```

Because `Symfony\Contracts\Service\ResetInterface` is also a dependency of another package in `require:` so it has been prefixed, moved and deleted, but `symfony/console/Application.php` still references the original because its part of a `require-dev:` dependency.

**Comment by BrianHenryIE** [comment]
I've started a branch which will probably take me months to finish. 

The idea is to track each file which is copied/deleted/prefixed and after changes have a list that can be queried. From that, I intend to build a file with a `class_alias(...)` list which should address the problem at the root of this issue – dev dependencies will look up the old namespace/classname and use the renamed one. There's a related `files` autoloader issue which this will help me fix too. Broadly, this will help move towards a Composer plugin structure where changes are all in-place and `composer install --no-dev` works smoothly too.

https://github.com/BrianHenryIE/strauss/compare/master...BrianHenryIE:strauss:use-files-objects-with-properties?expand=1

**Comment by defunctl** [comment]
@BrianHenryIE what do you think about adding a Strauss command line argument to configure `delete_vendor_packages` during execution?

Right now we get around this issue by setting that to `false`, but for a production zip, we'd like to delete the vendor packages, so like `strauss.phar --delete-vendor-packages` flag or option `strauss.phar --delete-vendor-packages=true|false` or something along those lines we could run to bypass the composer.json config.

We are just changing with `jq` for the time being, but an arg/option would be cleaner.

**Comment by BrianHenryIE** [comment]
Yeah, absolutely. I only recently added [a first CLI option](https://github.com/BrianHenryIE/strauss/blob/ccac6afa6d46b278911872e5e4c00c5b6004675f/src/Console/Commands/Compose.php#L55-L60) and [started writing progress to the terminal](https://github.com/BrianHenryIE/strauss/blob/ccac6afa6d46b278911872e5e4c00c5b6004675f/src/Console/Commands/Compose.php#L120). I'll try add that soon. It should be easy enough to add, but probably a pain to add a unit test given the current state of that class! I did look into it a bit before and there was something weird about handling both a `--delete-vendor-packages` and `--delete-vendor-packages=true`, which IMO should both work.

**Comment by BrianHenryIE** [comment]
OK, new feature, undocumented except in release notes, `vendor/bin/strauss --deleteVendorPackages=false` – https://github.com/BrianHenryIE/strauss/commit/73d2d3e1e8bd2d73d53e489f7dd19506a6dcbf14

Overrides the `composer.json` configuration.

Still plenty to do on this issue.

**Comment by defunctl** [comment]
@BrianHenryIE finally had a chance to test this, working well. Thank you!

**Comment by BrianHenryIE** [comment]
I finally got started on this. 

https://github.com/BrianHenryIE/strauss/compare/master...BrianHenryIE:strauss:add/alias-renamed-classes-for-dev-dependencies?expand=1

It has one passing test for `new \Psr\Log\NullLogger();` after renaming.

So far I'm creating a `autoload_renamed.php` file with aliases for classes, interfaces and traits. I need to get that to be loaded by Composer. When Strauss is loaded by Composer, a `files` autoloader for Strauss that loads a `bootstrap.php` which does a `file_exists()` on the new file might work, although it probably needs to be loaded after everything else. It'll need a different solution for when the `.phar` is used to perform the replacements. `classmap-authoritative` must be true for the code I've written to pick up the classes to alias, that could probably be programmatically created instead of relying on the `composer.json`. Hopefully I can eliminate the need to run `composer dump-autoload` after too.

**Comment by BrianHenryIE** [comment]
Complete in [0.22.0](https://github.com/BrianHenryIE/strauss/releases/tag/0.22.0), thank you. (report bugs!)
